XenAdmin/Diagnostics/Checks/DmcCheck.cs

```diff
@@ -0,0 +1,103 @@
+﻿/* Copyright (c) Citrix Systems, Inc. 
+ * All rights reserved. 
+ * 
+ * Redistribution and use in source and binary forms, 
+ * with or without modification, are permitted provided 
+ * that the following conditions are met: 
+ * 
+ * *   Redistributions of source code must retain the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer. 
+ * *   Redistributions in binary form must reproduce the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer in the documentation and/or other 
+ *     materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Windows.Forms;
+using XenAdmin.Core;
+using XenAdmin.Diagnostics.Hotfixing;
+using XenAdmin.Diagnostics.Problems;
+using XenAdmin.Diagnostics.Problems.HostProblem;
+using XenAdmin.Diagnostics.Problems.PoolProblem;
+using XenAPI;
+
+namespace XenAdmin.Diagnostics.Checks
+{
+    class DmcCheck : PoolCheck
+    {
+        private readonly Pool _pool;
+        private readonly Dictionary<string, string> _installMethodConfig;
+        private readonly bool _manualUpgrade;
+        private readonly Control _control;
+
+        public DmcCheck(Control control, Pool pool, Dictionary<string, string> installMethodConfig, bool manualUpgrade)
+            : base(pool)
+        {
+            _control = control;
+            _pool = pool;
+            _installMethodConfig = installMethodConfig;
+            _manualUpgrade = manualUpgrade;
+        }
+
+        public override bool CanRun()
+        {
+            return !Helpers.Post82X(_pool.Connection) && !Helpers.FeatureForbidden(_pool.Connection, Host.RestrictDMC);
+        }
+
+        protected override Problem RunCheck()
+        {
+            var coordinator = Helpers.GetCoordinator(Pool.Connection);
+
+            if (!_manualUpgrade)
+            {
+                var hotfix = HotfixFactory.Hotfix(coordinator);
+                if (hotfix != null && hotfix.ShouldBeAppliedTo(coordinator))
+                    return new HostDoesNotHaveHotfixWarning(this, coordinator);

```

Comment:
Question:
Why are we not also checking if supporter hosts need an hotfix?

Reply:
Further down Host.TryGetUpgradeVersion needs one host as a parameter, so the coordinator will do. Note though that when the RPU hotfix is applied, it is installed on the supporters too.

Reply:
Could there be an instance where the coordinator has a hotfix, but not the supporters?

Reply:
Now that I'm thinking about it, this check derives from PoolCheck, not HostCheck - I need to double check as I'm not sure any more whether it will be applied to all hosts. If there are further down HostChecks that need it though, it will be flagged by the pre-checks.

Reply:
I had a look at the code again: so this is a waring, which means it does not provide an action to install it. However, there is the  HostHasHotfixCheck at the beginning of the prechecks, which will show a blocking error for all the hosts to be upgraded and will offer the installation action. After that, the only way we could end up with a supporter not having the hotfix would be if installation failed or if a supporter was added to the pool after installation.

Reply:
Sounds good, Thank you for looking into it! <img src='https://emojis.slackmojis.com/emojis/images/1643514318/2905/thankyou.gif?1643514318' height='20' />